docs/specs/buffer.md

```diff
@@ -0,0 +1,31 @@
+# Buffer Specification
+
+This document specifies Vector's buffer behavior for the development of Vector.
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”,
+“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be
+interpreted as described in [RFC 2119].
+
+<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->
+
+1. [Instrumentation](#instrumentation)
+
+<!-- /MarkdownTOC -->
+
+## Instrumentation
+
+### When a Vector event is stored in the buffer
+* Metric
+  * MUST increment the `buffer_event_count` gauge by one
```

Comment:
I think we should consider the same [`received_events_total` metric](https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#eventsrecevied) that we use in components. Alignment with component metrics will reduce the cardinality. We can use tags to differentiate them:

```
received_events_total{entity=buffer} ...
```

cc @jszwedko on this. If we like this pattern we should probably update the component spec with the appropriate labels, `entity` specifically.

Reply:
Sure, that makes sense. We leaned toward the gauge initially as opposed to two counters (received_total, flushed_total) since it's a single metric and avoids having users derive the current buffer event count/bytes size. Happy to go in either direction.

Reply:
Yeah, we did discuss this. My thought was that, for buffers, users are generally going to care the most about the number of events currently in the buffer rather than the events flowing in and out. I think they differ from components in this respect since components can turn 0-n events into 0-n events such that knowing in/out is more useful.